src/util/Scheduler.cpp

```diff
@@ -0,0 +1,266 @@
+            {
+                mActions.pop_back();
+                return 1;
+            }
+        }
+        return 0;
+    }
+
+    void
+    enqueue(Action&& action, Scheduler::RelativeDeadline deadline)
+    {
+        auto elt = Element(std::move(action), deadline);
+        mActions.emplace_back(std::move(elt));
+    }
+
+    void
+    runNext(nsecs minServiceTime)
+    {
+        auto before = std::chrono::steady_clock::now();
+        Action action = std::move(mActions.front().mAction);
+        mActions.pop_front();
+        action();
+        auto after = std::chrono::steady_clock::now();
+        nsecs duration = std::chrono::duration_cast<nsecs>(after - before);
+        mServiceTime = std::max(mServiceTime + duration, minServiceTime);
+    }
+};
+
+Scheduler::Scheduler(size_t loadLimit,
+                     std::chrono::nanoseconds serviceTimeWindow)
+    : mQueueQueue(
+          [](std::shared_ptr<Queue> a, std::shared_ptr<Queue> b) -> bool {
+              return a->serviceTime() > b->serviceTime();
+          })
+    , mLoadLimit(loadLimit)
+    , mServiceTimeWindow(serviceTimeWindow)
+{
+}
+
+void
+Scheduler::trim(std::shared_ptr<Queue> q)
+{
+    Scheduler::AbsoluteDeadline now = std::chrono::steady_clock::now();
+    while (true)
+    {
+        size_t trimmed = q->tryTrim(mLoadLimit, now, mStats);
+        if (trimmed == 0)
+        {
+            return;
+        }
+        mSize -= trimmed;
+    }
+}
+
+void
+Scheduler::enqueue(std::string&& name, Action&& action,
+                   Scheduler::RelativeDeadline deadline)
+{
+    auto qi = mQueues.find(name);
+    if (qi == mQueues.end())
+    {
+        if (mQueueCache.exists(name))
+        {
+            mStats.mQueuesActivatedFromCache++;
+            qi = mQueues.emplace(name, mQueueCache.get(name)).first;
+        }
+        else
+        {
+            mStats.mQueuesActivatedFromFresh++;
+            auto q = std::make_shared<Queue>(name);
+            mQueueCache.put(name, q);
+            qi = mQueues.emplace(name, q).first;
+        }
+        mQueueQueue.push(qi->second);
+    }
+    mStats.mActionsEnqueued++;
+    qi->second->enqueue(std::move(action), deadline);
+    mSize += 1;
+    trim(qi->second);
+}
+
+size_t
+Scheduler::runOne()
+{
+    if (mQueueQueue.empty())
+    {
+        return 0;
+    }
+    else
+    {
+        auto q = mQueueQueue.top();
+        mQueueQueue.pop();
+        trim(q);
+        if (!q->isEmpty())
+        {
+            // We pass along a "minimum service time" floor that the service
+            // time of the queue will be incremented to, at minimum.
+            auto minServiceTime = mMaxServiceTime - mServiceTimeWindow;
+            q->runNext(minServiceTime);
+            mMaxServiceTime = std::max(q->serviceTime(), mMaxServiceTime);
+            mSize -= 1;
+            mStats.mActionsDequeued++;
+            trim(q);
+        }
+        if (q->isEmpty())
```

Comment:
Fixed thought differently (see my other comment -- it can't be in the runnable queue). I added a list-based secondary idle queue with expiry from the end.